### PR TITLE
Issue #8901: Run regression testing in CI over openjdk14 or newer codebase

### PR DESCRIPTION
**Pull request by nmancus1**
Issue #8901: Run regression testing in CI over openjdk14 or newer codebase

This PR will allow us to parse the entire(almost) openjdk14u repo; all excluded files are listed in `openjdk14-excluded.files` in the form of `BeforeExecutionExclusionFileFilter` modules that are appended to the `checks-nonjavadoc-error.xml` config file via `sed` when running Checkstyle over the repo.

TODO list:
- [x] After https://github.com/checkstyle/contribution/pull/516 is merged, update `openjdk14-with-checks-nonjavadoc-error` to have the correct contribution repo link

- [x] Create or find applicable issues for excluded files:

    - [x] `jdk14u/test/jdk/sun/tools/jhsdb/HeapDumpTest.java`
    - [x] `jdk14u/test/jdk/java/lang/String/TranslateEscapes.java`
    - [x] `jdk14u/test/hotspot/jtreg/runtime/exceptionMsgs/methodPrinting/TestPrintingMethods.java`
    - [x] `jdk14u/test/langtools/tools/javac/annotations/typeAnnotations/failures/T8011722.java`
    - [x] `jdk14u/test/langtools/tools/javac/annotations/typeAnnotations/referenceinfos/NestedTypes.java`
    - [x] `jdk14u/test/langtools/tools/javac/annotations/typeAnnotations/newlocations/AllLocations.java`
    - [x] `jdk14u/test/langtools/tools/javac/enum/TrailingComma.java`
    - [x] `jdk14u/test/langtools/tools/javac/treeannotests/AnnoTreeTests.java`
    - [x] `jdk14u/test/langtools/tools/javac/lambda/speculative/InferStrict.java`
    - [x] `jdk14u/test/langtools/tools/javac/DeepStringConcat.java`
    - [x] `jdk14u/test/langtools/tools/javac/switchexpr/ExpressionSwitchBugs.java`  

<br/>

- [x] Create issue to disable/remove `checkstyle-cli-run-openjdk14` in `travis.sh`, as it is superseded by this test.  `checkstyle-cli-run-openjdk14` parsed around 2,000 files; and this new test parses around 46,000 files(this excludes module-info.java files and non-compilables):

```
➜  jdk14u git:(master) find . -name "*.java" | wc -l
49160

```

### CODE REVIEW

File: .ci/no-exception-test.sh [hunk 1]
```diff
@@ -53,6 +53,35 @@ guava-with-sun-checks)
...
+openjdk14-with-checks-nonjavadoc-error)
+  CS_POM_VERSION=$(mvn -e -q -Dexec.executable='echo' -Dexec.args='${project.version}' \
+                     --non-recursive org.codehaus.mojo:exec-maven-plugin:1.3.1:exec)
+  echo CS_version: $CS_POM_VERSION
+  mkdir -p .ci-temp/
+  cd .ci-temp/
+  # activate below line after https://github.com/checkstyle/contribution/pull/516
+  # git clone https://github.com/checkstyle/contribution
+  # need to remove below line after https://github.com/checkstyle/contribution/pull/516
+  git clone https://github.com/nmancus1/contribution.git
```
Comment (nmancus1) [inline .ci/no-exception-test.sh:65]: Remaining travis failure is due to spellcheck failing on "nmancus", which will be removed after https://github.com/checkstyle/contribution/pull/516

**Comment by nmancus1** [comment]
Passing semaphore build: https://checkstyle.semaphoreci.com/jobs/f0589693-fcba-4781-b12e-e32d2ae2f225
image

**Comment by nmancus1** [comment]
@romani ping

File: .ci/openjdk-projects-to-test-on.properties [hunk 2]
```diff
@@ -0,0 +1,11 @@
+# suppress inspection "UnusedProperty" for whole file
+# suppress inspection "WrongPropertyKeyValueDelimiter" for whole file
```
Comment (romani) [inline .ci/openjdk-projects-to-test-on.properties:2]: please change extension of file to ".config" to skip this file from any tool that try to validate it.

I do not remember why we named it as ".properties", probably very first version of file was based on properties.

File: .ci/jsoref-spellchecker/whitelist.words [hunk 3]
```diff
@@ -1347,6 +1348,7 @@ tt
 ttf
 ttype
 tw
+Twr
```
Comment (romani) [inline .ci/jsoref-spellchecker/whitelist.words:1351]: bunch of crappy words are leaking to our whitelist, lets exclude .ci/openjdk14-excluded.files completely from spell-checker

File: .ci/no-exception-test.sh [hunk 4]
```diff
@@ -53,6 +53,28 @@ guava-with-sun-checks)
...
 
+openjdk14-with-checks-nonjavadoc-error)
+  CS_POM_VERSION=$(mvn -e -q -Dexec.executable='echo' -Dexec.args='${project.version}' \
+                     --non-recursive org.codehaus.mojo:exec-maven-plugin:1.3.1:exec)
+  echo CS_version: $CS_POM_VERSION
+  mkdir -p .ci-temp/
+  cd .ci-temp/
+  git clone https://github.com/checkstyle/contribution
+  cd ..
+  mvn -e clean install -Pno-validations
```
Comment (romani) [inline .ci/no-exception-test.sh:64]: please move this to build line to be done before clone of contribution

**Review by romani — CHANGES_REQUESTED** [review]
items:

File: .ci/no-exception-test.sh [hunk 5]
```diff
@@ -53,6 +53,28 @@ guava-with-sun-checks)
...
+  sed -i.'' 's/value=\"error\"/value=\"ignore\"/' \
+        .ci-temp/contribution/checkstyle-tester/checks-nonjavadoc-error.xml
+  cd .ci-temp/contribution/checkstyle-tester
+  sed -i '/  <!-- Filters -->/r ../../../.ci/openjdk14-excluded.files' checks-nonjavadoc-error.xml
```
Comment (romani) [inline .ci/no-exception-test.sh:68]: awesome hack ! I like it.
I would keep it in mind, it might be handy for other complicated cases also.

**Comment by romani** [comment]
> ExpressionSwitchBugs.java

is it case what we did not noticed during summer ?

File: .ci/no-exception-test.sh (hunk 5 above)
Comment (nmancus1) [reply .ci/no-exception-test.sh:68]: I'll leave this here for the future then...
Ran from Checkstyle root:

```
#!/bin/bash
OUTPUT="test"
while  [ -n "$OUTPUT" ]
do
    OUTPUT=$(.ci/no-exception-test.sh openjdk14-checkstyle-checks-no-exception | grep "Caused by: " | grep -oh '/.*.java' | sed 's|.*/||' | cut -f 1 -d '.' | uniq)
    echo "$OUTPUT"
    echo -e "<module name=\"BeforeExecutionExclusionFileFilter\">\n  <property name=\"fileNamePattern\" value=\"/$OUTPUT\.java\$\"/>\n</module>" >> config/openjdk14-excluded.files
    rm -rf .ci-temp/contribution
done
```

**Comment by nmancus1** [comment]
>is it case what we did not noticed during summer ?

Seems like this shouldn't be too hard to fix:

https://github.com/checkstyle/checkstyle/issues/8934

**Comment by nmancus1** [comment]
The three remaining files in the above checklist, `jdk14u/test/langtools/tools/javac/annotations/typeAnnotations/referenceinfos/NestedTypes.java`,
 `jdk14u/test/langtools/tools/javac/policy/test1/C.java`, and
 `jdk14u/test/langtools/tools/javac/lambda/speculative/InferStrict.java`

Seem to intermittently fail.  I am trying to reproduce this failure to open a bug report now.  All other files have been associated with open issues, or a new issue opened for them.

**Comment by nmancus1** [comment]
All excluded files have either had issues opened for them, or been associated with open issues within the `openjdk14-excluded.files` file.

File: .ci/openjdk-projects-to-test-on.properties (hunk 2 above)
Comment (nmancus1) [reply .ci/openjdk-projects-to-test-on.properties]: Done.

File: .ci/jsoref-spellchecker/whitelist.words (hunk 3 above)
Comment (nmancus1) [reply .ci/jsoref-spellchecker/whitelist.words]: Done.

File: .ci/no-exception-test.sh (hunk 4 above)
Comment (nmancus1) [reply .ci/no-exception-test.sh]: Done.

File: .ci/no-exception-test.sh (hunk 5 above)
Comment (nmancus1) [reply .ci/no-exception-test.sh:68]: I've linked to the above script from the generated file in this PR, in case we should need to create such a file in the future.

File: .ci/no-exception-test.sh (hunk 4 above)
Comment (nmancus1) [reply .ci/no-exception-test.sh]: Just curious, why? In case of build failure, we don't waste resources/time to clone the contribution repo?

**Comment by nmancus1** [comment]
Files/modules that can be removed from `openjdk14-excluded.files` are related to 
https://github.com/checkstyle/checkstyle/issues/8945
https://github.com/checkstyle/checkstyle/issues/8934
https://github.com/checkstyle/checkstyle/issues/8656
https://github.com/checkstyle/checkstyle/issues/8654
https://github.com/checkstyle/checkstyle/issues/8650
https://github.com/checkstyle/checkstyle/issues/8651
https://github.com/checkstyle/checkstyle/issues/8655
https://github.com/checkstyle/checkstyle/issues/8291
https://github.com/checkstyle/checkstyle/issues/8652
https://github.com/checkstyle/checkstyle/issues/8653
https://github.com/checkstyle/checkstyle/issues/8943
I'm putting this here for visibility when those issues are closed.

File: .ci/no-exception-test.sh (hunk 4 above)
Comment (romani) [reply .ci/no-exception-test.sh]: build failure is rare
in CI we do not make waste, as whole environment is deleted after al executions, only logs are left.
on local, yes it might, but usually this command executed the last verification, so build is passing for sure.

File: .ci/openjdk14-excluded.files [hunk 6]
```diff
@@ -0,0 +1,514 @@
+<!-- This file generated by script at -->
+<!--     https://github.com/checkstyle/checkstyle/pull/8920#discussion_r513557040 -->
```
Comment (romani) [inline .ci/openjdk14-excluded.files:2]: please make it as comment in issue, it is more visible area.

all PR conversations become resolved "collabsed".

File: .semaphore/semaphore.yml [hunk 7]
```diff
@@ -84,6 +84,7 @@ blocks:
               values:
                 - guava-with-google-checks
                 - guava-with-sun-checks
+                - openjdk14-with-checks-nonjavadoc-error
```
Comment (romani) [inline .semaphore/semaphore.yml:87]: @pbludov , this is not very good approach for CI config.

it should be:
`.ci/no-exception-test.sh openjdk14-with-checks-nonjavadoc-error`

and in comamnds we do `eval $COMMAND`

one day this execution will fail in CI in PR, contributor might not be very experienced in CI to figureout how to reproduce it on local, so we will share with him link to single line in config for simple copy and paste in terminal, and it works.

**Review by romani — APPROVED** [review]
@nmancus1 , as always ... awesome work, thanks a lot.

please update all issue to make note to remove related filter after fix.

File: .ci/no-exception-test.sh (hunk 5 above)
Comment (romani) [reply .ci/no-exception-test.sh:68]: @nmancus1 , see my comment below on how to improve.

File: .semaphore/semaphore.yml (hunk 7 above)
Comment (pbludov) [reply .semaphore/semaphore.yml:87]: done in https://github.com/checkstyle/checkstyle/pull/8950

File: .ci/no-exception-test.sh [hunk 8]
```diff
@@ -53,6 +53,28 @@ guava-with-sun-checks)
   rm sun_checks.*
   ;;
 
+openjdk14-with-checks-nonjavadoc-error)
+  CS_POM_VERSION=$(mvn -e -q -Dexec.executable='echo' -Dexec.args='${project.version}' \
+                     --non-recursive org.codehaus.mojo:exec-maven-plugin:1.3.1:exec)
+  echo CS_version: $CS_POM_VERSION
+  mvn -e clean install -Pno-validations
```
Comment (pbludov) [inline .ci/no-exception-test.sh:60]: Checkstyle is built already: https://github.com/checkstyle/checkstyle/blob/master/.semaphore/semaphore.yml#L88

Comment (pbludov) [reply .ci/no-exception-test.sh:60]: i'll remove duplicate from swmaphore.yml

**Review by pbludov — APPROVED** [review]

File: .ci/no-exception-test.sh (hunk 8 above)
Comment (romani) [reply .ci/no-exception-test.sh:60]: this is a prat I still do not have strong position, where to keep build of checkstyle in validation commands or in CI config (presumably it is "given").
Some CI reuse the same host for all executions (wercker for example).
probably we can have shells like:
`ci.yml -> your-ci.sh TASK_NAME -> "mvn install ...."  && validation.sh TASKNAME `

Comment (nmancus1) [reply .ci/no-exception-test.sh:60]: From above:
>one day this execution will fail in CI in PR, contributor might not be very experienced in CI to figureout how to reproduce it on local, so we will share with him link to single line in config for simple copy and paste in terminal, and it works.

If we offload the building of checkstyle to CI/host and remove it from script, it will not be a 'single line in config for simple copy and paste in terminal' anymore.

Comment (romani) [reply .ci/no-exception-test.sh:60]: in my case it is simple copy-paste with presumption that sources are built.
if no built done, failure is obvious
when  no re-build is done, significant execution time reduction.

**Comment by nmancus1** [comment]
I'm going to hold off on this PR until https://github.com/checkstyle/checkstyle/pull/8950 so that I can incorporate whatever changes/decisions we make to CI execution into this PR, if that's ok.

**Comment by romani** [comment]
@nmancus1 , please continue. Referenced PR is merged

**Comment by nmancus1** [comment]
>please update all issue to make note to remove related filter after fix.

Done.

**Comment by nmancus1** [comment]
I wasn't sure whether to use the "fast pool" or "slow pool" section, I will compare runtimes after completion. I suspect this test should be in the "slow pool" group, though.

File: .semaphore/semaphore.yml [hunk 9]
```diff
@@ -73,6 +73,8 @@ blocks:
                     && .ci/validation.sh no-error-pmd
                 - mvn -e clean install -Pno-validations
                     && .ci/validation.sh no-violation-test-configurate
+                - mvn -e clean install -Pno-validations
+                  && .ci/no-exception-test.sh openjdk14-with-checks-nonjavadoc-error
```
Comment (pbludov) [inline .semaphore/semaphore.yml:77]: please add indentation for easy reading

**Review by pbludov — CHANGES_REQUESTED** [review]
Just one minor change

File: .semaphore/semaphore.yml (hunk 9 above)
Comment (nmancus1) [reply .semaphore/semaphore.yml:77]: Fixed.

File: .ci/openjdk14-excluded.files (hunk 6 above)
Comment (nmancus1) [reply .ci/openjdk14-excluded.files]: Done.

**Review by pbludov — APPROVED** [review]

File: .ci/openjdk14-excluded.files [hunk 10]
```diff
@@ -0,0 +1,514 @@
+<!-- This file generated by script at -->
+<!--     https://github.com/checkstyle/checkstyle/issues/8901#issuecomment-712939037 -->
+
+<!-- Excluded due to exception; issues opened -->
+
+<!-- until https://github.com/checkstyle/checkstyle/issues/8945 -->
+<module name="BeforeExecutionExclusionFileFilter">
+  <property name="fileNamePattern" value="/NestedTypes\.java$"/>
```
Comment (rnveach) [inline .ci/openjdk14-excluded.files:8]: Is there a reason these files do not have any reference to an exact or matching folder path?

By specifying a file only, it means we will exclude any file matching that name, especially if there are 2 or more.

Furthermore why do some have `/` and others don't? For Windows compatibility folder paths should be `[\\/]`

File: .ci/openjdk-projects-to-test-on.config [hunk 11]
```diff
@@ -0,0 +1,8 @@
+# List of GIT repositories to clone / pull for checking with Checkstyle
+# File format: REPO_NAME|[local|git|hg]|URL|[COMMIT_ID]|[EXCLUDE FOLDERS]
+# Please note that bash comments works in this file
+#
+# This file has been created to alleviate confusion in the contribution
+# repository's projects-to-test-on.properties file; these projects
+# must be used with special configurations.
+openjdk14|git|https://github.com/openjdk/jdk14|master||
```
Comment (rnveach) [inline .ci/openjdk-projects-to-test-on.config:8]: Why is this file needed? Why can't we use contribution's file and tell it to target openjdk14 only?

File: .ci/no-exception-test.sh [hunk 12]
```diff
@@ -53,6 +53,27 @@ guava-with-sun-checks)
...
+  mkdir -p .ci-temp/
+  cd .ci-temp/
+  git clone https://github.com/checkstyle/contribution
+  cd ..
+  sed -i.'' 's/value=\"error\"/value=\"ignore\"/' \
+        .ci-temp/contribution/checkstyle-tester/checks-nonjavadoc-error.xml
+  cd .ci-temp/contribution/checkstyle-tester
+  sed -i '/  <!-- Filters -->/r ../../../.ci/openjdk14-excluded.files' checks-nonjavadoc-error.xml
+  export MAVEN_OPTS="-Xmx2048m"
+  groovy ./launch.groovy --listOfProjects ../../../.ci/openjdk-projects-to-test-on.config \
```
Comment (rnveach) [inline .ci/no-exception-test.sh:69]: @romani Note this is using launch and not diff.
Not expecting a change here.

File: .ci/openjdk-projects-to-test-on.config (hunk 11 above)
Comment (romani) [reply .ci/openjdk-projects-to-test-on.config:8]: It is very special line, it  is not usable for regular groovy files.
I even removed openjfk from contribution repo, as it caused too much confusion to contributors.

File: .ci/openjdk14-excluded.files (hunk 10 above)
Comment (romani) [reply .ci/openjdk14-excluded.files:8]: https://github.com/checkstyle/checkstyle/issues/8901#issuecomment-712939037

It was easier to catch files and generate this file.
Change to match different files is very low. We can improve this late on ones it become proven way for jdk validation.

File: .ci/no-exception-test.sh (hunk 12 above)
Comment (romani) [reply .ci/no-exception-test.sh:69]: @nmancus1 , can we use diff.grovy here ?

Comment (nmancus1) [reply .ci/no-exception-test.sh:69]: I can add the same `extraMvnOptions` command line arg to to `diff.groovy`, but since `diff.groovy` uses `launch.groovy` anyway, why not keep it as is and not have another utility to deal with?  Just curious.

File: .ci/openjdk14-excluded.files (hunk 10 above)
Comment (rnveach) [reply .ci/openjdk14-excluded.files:8]: I understand this file is generated, but I am seeing anything that says why folders were removed. If this is to be done later ok, but we could be missing overlapping files.

File: .ci/openjdk-projects-to-test-on.config (hunk 11 above)
Comment (rnveach) [reply .ci/openjdk-projects-to-test-on.config:8]: I am resolving this. I thought the line was still in contribution.

File: .ci/openjdk14-excluded.files (hunk 10 above)
Comment (nmancus1) [reply .ci/openjdk14-excluded.files:8]: >Is there a reason these files do not have any reference to an exact or matching folder path?

A good question. There were quite a few duplicates of identical/ nearly identical files with the same name that kept failing with NPE, so in order to generate this list of 
 `BeforeExecutionExclusionFileFilter` modules in a reasonable amount of time (it still took over three days),  I had the script eliminate the path. 

Also, since I semi-manually verified non-compilable files, this allowed me to get it done in less time. Additionally, I have had issues line-wrapping XML reliably, so another reason to omit the paths was to keep lines under 100 chars.

>By specifying a file only, it means we will exclude any file matching that name, especially if there are 2 or more in different folders.

I felt that losing some granularity in the `BeforeExecutionExclusionFileFilter` modules was a fair trade to be able to increase our testing on openjdk14 from around 2,000 files to more than 40,000.  In retrospect, I should have discussed this in issue and got feedback before proceeding. 

>Furthermore why do some have / and others don't?
That's a good question; I didn't catch that.  The file names that are not preceded with '/' were the ones that I added manually that we were already aware of issues with.  I can fix that.

>For Windows compatibility folder paths should be `[\\/]`

We only have a Linux build on Semaphore, so I hadn't considered Windows compatibility, honestly.  I can fix this so that windows users can run locally.

File: .ci/no-exception-test.sh (hunk 12 above)
Comment (rnveach) [reply .ci/no-exception-test.sh:69]: This goes back to discussion in main issue, ending with https://github.com/checkstyle/checkstyle/issues/8901#issuecomment-717596429 .

If we are to remove JXR from tester's POM to avoid telling it not to run anyways, we need to remove uses of `launch` as I think we are still relying on it generating HTML files since diff-report-tool is what generates the final HTML when `diff` is launched.

Comment (nmancus1) [reply .ci/no-exception-test.sh:69]: So,  I should submit a PR to contribution to add `extraMvnOptions` to diff.groovy, then use that in this PR, right?

File: .ci/openjdk14-excluded.files (hunk 10 above)
Comment (romani) [reply .ci/openjdk14-excluded.files:8]: Please fix windows compatibility.
Another issue can be postponed.

File: .ci/no-exception-test.sh (hunk 12 above)
Comment (romani) [reply .ci/no-exception-test.sh:69]: I recommend to merge now and improve to use diff.groovy in next iteration/PR.

File: .ci/openjdk14-excluded.files (hunk 10 above)
Comment (rnveach) [reply .ci/openjdk14-excluded.files:8]: Yea, please fix Windows directories so it can be run by those users locally.

I am ok to skip adding any type of folder paths.

File: .ci/no-exception-test.sh (hunk 12 above)
Comment (rnveach) [reply .ci/no-exception-test.sh:69]: Then this is resolved for this PR, I will just leave this open for follow up.

Comment (romani) [reply .ci/no-exception-test.sh:69]: @nmancus1 , please create issue on this , one day we will address it.

**Comment by romani** [comment]
There is PR for new syntax ... 
I need this PR be merged sooner to start validation in PRs validation.

File: .ci/openjdk14-excluded.files (hunk 10 above)
Comment (nmancus1) [reply .ci/openjdk14-excluded.files:8]: To be clear, I need to change all the preceding `/` to `\\/`, right?

File: .ci/no-exception-test.sh (hunk 12 above)
Comment (nmancus1) [reply .ci/no-exception-test.sh:69]: https://github.com/checkstyle/checkstyle/issues/8960
https://github.com/checkstyle/contribution/issues/523

File: .ci/openjdk14-excluded.files (hunk 10 above)
Comment (romani) [reply .ci/openjdk14-excluded.files:8]: you have single , so need to change single.
example of how it will looks like:
https://github.com/checkstyle/checkstyle/blob/3fc9dd98ad6ed963e2cfcd2e630f7031ec5b0c1e/config/suppressions.xml#L67

Comment (nmancus1) [reply .ci/openjdk14-excluded.files]: Fixed.  I've also removed to escapes from the `\.` in file names to be consistent with your example.

**Comment by romani** [comment]
looks good, as CI pass we will merge it.
@nmancus1 , thanks a lot for your help.
